### PR TITLE
Fix stored coefficients to make M2L faster for weird kernels

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

File: sumpy/expansion/__init__.py [hunk 1]
```diff
@@ -282,12 +282,11 @@ def _split_coeffs_into_hyperplanes(self) -> List[Tuple[int, List[Tuple[int]]]]:
...
                 # Calculate the multi-index that appears last in in the PDE in
-                # reverse degree lexicographic order (degrevlex).
-                # The monomial ordering used here which is degrevlex, must match
-                # the monomial ordering used in LinearPDEBasedExpansionTermsWrangler
-                max_mi_idx = max(mi_to_index[ident.mi] for
-                                 ident in pde_dict.keys())
-                max_mi = mis[max_mi_idx]
+                # degree lexicographic order.
+                ordering_key = self._get_mi_ordering_key()
+                pde_dict, = self.knl.get_pde_as_diff_op().eqs
```
Comment (inducer) [inline sumpy/expansion/__init__.py:287]: - The same line is just above. Does it need to be repeated?
- Renaming `pde_dict` to `deriv_id_to_coeff` would be easier to read IMO.

File: sumpy/expansion/__init__.py [hunk 2]
```diff
@@ -282,12 +282,11 @@ def _split_coeffs_into_hyperplanes(self) -> List[Tuple[int, List[Tuple[int]]]]:
...
             else:
                 # Calculate the multi-index that appears last in in the PDE in
-                # reverse degree lexicographic order (degrevlex).
-                # The monomial ordering used here which is degrevlex, must match
-                # the monomial ordering used in LinearPDEBasedExpansionTermsWrangler
-                max_mi_idx = max(mi_to_index[ident.mi] for
-                                 ident in pde_dict.keys())
-                max_mi = mis[max_mi_idx]
+                # degree lexicographic order.
+                ordering_key = self._get_mi_ordering_key()
```
Comment (inducer) [inline sumpy/expansion/__init__.py:286]: It looks like `_get_mi_ordering_key` is only defined in `LinearPDEBasedExpansionTermsWrangler`. Wouldn't this make `ExpansionTermsWrangler._split_coeffs_into_hyperplanes` unusable unless we're in that subclass? Maybe this wants to be a function instead?

File: sumpy/expansion/__init__.py [hunk 3]
```diff
@@ -442,6 +441,43 @@ def stored_identifiers(self):
         stored_identifiers, _ = self.get_stored_ids_and_unscaled_projection_matrix()
         return stored_identifiers
 
+    @memoize_method
+    def _get_mi_ordering_key(self):
+        """Calculate the multi-index that appears last in the PDE given the pde_dict
+        A degree lexicographic order with the slowest varying index depending on
+        the PDE is used. For two dimensions, this is either deglex or degrevlex.
```
Comment (inducer) [inline sumpy/expansion/__init__.py:448]: - Grammar?
- Explain this a bit (in a comment)?
- What's the impact?

File: sumpy/expansion/__init__.py [hunk 4]
```diff
@@ -442,6 +441,43 @@ def stored_identifiers(self):
...
+        """
+        dim = self.dim
+        pde_dict, = self.knl.get_pde_as_diff_op().eqs
```
Comment (inducer) [inline sumpy/expansion/__init__.py:451]: Rename `pde_dict` as above.

File: sumpy/expansion/__init__.py [hunk 5]
```diff
@@ -282,12 +282,11 @@ def _split_coeffs_into_hyperplanes(self) -> List[Tuple[int, List[Tuple[int]]]]:
...
-                # reverse degree lexicographic order (degrevlex).
-                # The monomial ordering used here which is degrevlex, must match
-                # the monomial ordering used in LinearPDEBasedExpansionTermsWrangler
-                max_mi_idx = max(mi_to_index[ident.mi] for
-                                 ident in pde_dict.keys())
-                max_mi = mis[max_mi_idx]
+                # degree lexicographic order.
+                ordering_key = self._get_mi_ordering_key()
+                pde_dict, = self.knl.get_pde_as_diff_op().eqs
+                max_mi = max((ident for ident in pde_dict.keys()),
```
Comment (inducer) [inline sumpy/expansion/__init__.py:288]:
```suggestion
                max_mi = max((ident for ident in pde_dict),
```

File: sumpy/expansion/__init__.py [hunk 6]
```diff
@@ -442,6 +441,43 @@ def stored_identifiers(self):
...
+    @memoize_method
+    def _get_mi_ordering_key(self):
+        """Calculate the multi-index that appears last in the PDE given the pde_dict
+        A degree lexicographic order with the slowest varying index depending on
+        the PDE is used. For two dimensions, this is either deglex or degrevlex.
+        """
+        dim = self.dim
+        pde_dict, = self.knl.get_pde_as_diff_op().eqs
+        slowest_varying_index = 0
+        for ident in pde_dict.keys():
```
Comment (inducer) [inline sumpy/expansion/__init__.py:453]:
```suggestion
        for ident in pde_dict:
```

File: sumpy/expansion/__init__.py [hunk 7]
```diff
@@ -442,6 +441,43 @@ def stored_identifiers(self):
...
+            if ident.mi.count(0) == dim - 1:
+                non_zero_index = next(i for i in range(self.dim) if ident.mi[i] != 0)
+                slowest_varying_index = max(slowest_varying_index, non_zero_index)
+
+        mi_compare_axis = list(range(self.dim))
+        mi_compare_axis[0], mi_compare_axis[slowest_varying_index] = \
+                slowest_varying_index, 0
```
Comment (inducer) [inline sumpy/expansion/__init__.py:460]: It looks like you're computing an axis permutation here. Maybe name it as such?

**Review by inducer — COMMENTED** [review]
Thanks for working on this! A few thoughts from a scroll through the code.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: sumpy/expansion/__init__.py (hunk 7 above)
Comment (isuruf) [reply sumpy/expansion/__init__.py]: Done in 90e77d0

File: sumpy/expansion/__init__.py (hunk 6 above)
Comment (isuruf) [reply sumpy/expansion/__init__.py]: Done in 1fa3e19

File: sumpy/expansion/__init__.py (hunk 5 above)
Comment (isuruf) [reply sumpy/expansion/__init__.py]: Done in 1fa3e19

File: sumpy/expansion/__init__.py (hunk 4 above)
Comment (isuruf) [reply sumpy/expansion/__init__.py]: Done in 4b26ae0

File: sumpy/expansion/__init__.py (hunk 1 above)
Comment (isuruf) [reply sumpy/expansion/__init__.py]: Removed the line as it was a duplicate.
Rename done in 4b26ae0

File: sumpy/expansion/__init__.py (hunk 2 above)
Comment (isuruf) [reply sumpy/expansion/__init__.py]: Fixed in 566afb7 by adding a method that the subclass overrides.

File: sumpy/expansion/__init__.py (hunk 3 above)
Comment (isuruf) [reply sumpy/expansion/__init__.py]: I explained a bit more in 827f067. (I also removed the for two dimensions line as it was unnecessary)

**Comment by isuruf** [comment]
@inducer, this is ready for a review.

File: sumpy/expansion/__init__.py [hunk 8]
```diff
@@ -241,6 +241,21 @@ def copy(self, **kwargs):
 
         return type(self)(**new_kwargs)
 
+    def _get_coeff_hyperplanes(self) -> List[Tuple[int, int]]:
```
Comment (inducer) [inline sumpy/expansion/__init__.py:244]:
```suggestion
    def _get_multiindex_hyperplanes(self) -> List[Tuple[int, int]]:
```

File: sumpy/expansion/__init__.py [hunk 9]
```diff
@@ -442,6 +428,68 @@ def stored_identifiers(self):
         stored_identifiers, _ = self.get_stored_ids_and_unscaled_projection_matrix()
         return stored_identifiers
 
+    @memoize_method
+    def _get_mi_ordering_key(self):
+        """Calculate the multi-index that appears last in the PDE.
```
Comment (isuruf) [inline sumpy/expansion/__init__.py:433]:
```suggestion
        """
```

**Review by inducer — APPROVED** [review]
Thanks! LGTM with these comment/doc changes.

**Comment by inducer** [comment]
LGTM, thanks! (Assuming you meant for this to be ready to review but forgot to rerequest.)

**Comment by isuruf** [comment]
Thanks

**Pull request merged**